### DISCUSSION TITLE
Improve container rebuild performance, features installation order and pinning version

### DISCUSSION THREAD

**Discussion by audacioustux** [devcontainers/spec#327] [Ideas]
the problem I've in hand is, devcontainer build takes a long time, Codespaces prebuild is not the solution I'm looking for.
rebuilding is something very common, as I try to modularize [configurations as features](https://github.com/audacioustux/devcontainers) frequently. the `features` entry in `devcontainers.json` is an object, so there's not an explicit agreement of the order, and based on `installsAfter` config in each feature. That means the new feature added may not be the last feature in that order - and break the container build cache of many features needlessly.

one solution would be to `overrideFeatureInstallOrder`, but that becomes too tedious and error prone. I think it'd be more ideal if `features` was an array, that hints the preferred order, and then dictated by `installsAfter`. I don't think it's a new idea, as it's kinda obvious one to me. hope someone can point me to relevant discussion on this, or if I'm wrong in my assumptions.

another thing that would be nice, to have a button (alongside the `update to latest` button) that sets feature version to patch release. for example, if the version is set to `some-feature:1` convert that to `some-feature:1.0.2`.
![image](https://github.com/devcontainers/spec/assets/7238675/e6f498c1-1693-4085-aeeb-5f87b21081b6)
btw I think there's a problem here - the suggested version is lower than what already is set!

I know this repo is for only the spec, and some portion of what I've said is not directly related to the spec.

**Comment by joshspicer** [comment]
Hey there! To your comment on installation order, there's been a fair bit of discussion on this in past discussion (if you're curious on the historical reasoning https://github.com/devcontainers/spec/issues/43, https://github.com/devcontainers/spec/issues/109).  There's three ways to influence installation order (see https://github.com/devcontainers/spec/blob/main/docs/specs/devcontainer-features.md#installation-order).  Note that with `overrideFeatureInstallOrder` you don't need to list every Feature, just the one(s) you need "pulled to the top". That might make the usage of that property less tedious, although I understand that isn't what you're looking for and won't solve the caching problem.

Thanks for flagging the codelens issue and for the feedback - looks like a string sorting issue that i'll take a look at!
